UDS Background Section

SteveRyan-ASU opened the pull request:
This block establishes an optional `<section>` element exclusively to wrap a standard Bootstrap 4 container. The purpose of this wrapper is to:

- Provide a consistent semantic HTML5 markup for the page.
- Provide a consistent class to use to reinforce the max-width:1920px rule.
- Provide a block interface to select approved ASU background textures or approved colors from a list.
- Provide a block interface to allow users to upload their own background texture.
- Automatically set the margin and padding rules for this element in the context of surrounding elements.
- Automatically set the background color and text color on the interior container based on the supplied background texture.

The output of the block is:
```
<section class="uds-section ..." >
...
</section>
```
...where `uds-section` is followed by a series of optional utility classes. Other notes:

- It allows only two blocks as inner content: the Bootstrap container block and the custom HTML block.
- It also comes with sample content that pre-populates the `<section>` with an interior `<container>` when you first add the block to the editor.

SteveRyan-ASU commented:
Controls for the block via ACF. (Radio buttons all the way down.)
![image](https://user-images.githubusercontent.com/2085753/109263016-cefbcf00-77bf-11eb-84d7-0e5c18c40ea4.png)

SteveRyan-ASU commented:
A note on the usability of this block within the editor:

- The interface for using this block as the starting point for a new section still feels very clunky.
- It's relatively easy to set up the remaining BS row and columns incorrectly. The background colors and textures make it difficult to see where to click next for the row and columns.

It works well enough out of the box, but a better use case for this would be to establish a block pattern that added this block + the container/row/columns for common layouts.

SteveRyan-ASU commented:
If this block is merged it would close #86. 

While not exactly what was prescribed in that issue, this block + the other blocks added by the container block plugin get us to the intended end result.

SteveRyan-ASU commented:
One additional note: I was hoping that we would be able to drag a block (or a series of blocks) into this block like you can with the native media/text block (as an example.) But, I was unable to figure out how to do that within the ACF context.

zainabD reviewed (changes requested):
Good work Steve!, I noticed some issues when using the block with different combinations, also added some suggestions below:

- [ ] [Issue] The Additional CSS class(es) field is not connected.
- [Suggestion] We need to add the option to change the container's background to transparent, as in some cases we need to add text, header or button over the background section without having to have another colored background. 
![image](https://user-images.githubusercontent.com/15880606/109559542-099f8900-7a98-11eb-9f41-551c873df24f.png)

- [Question] Why can't we get the opacity options when selecting the UDS background texture, while these 3 options become available when uploading an image?
![image](https://user-images.githubusercontent.com/15880606/109562778-11612c80-7a9c-11eb-8dfc-a0a5d8a392fb.png)


- [ ] [Issue] When I check the **Fluid** option on the container block, the container background becomes transparent, since you are targeting the **.uds-section.bg.topo-black .container**, not .**container-fluid**. 
- [ ] [Issue] In dashboard view, when checking the **Fluid** option, the container block remains the same with background color as you're targeting .**editor-styles-wrapper .uds-section.bg.topo-black .wp-block-wp-bootstrap-blocks-container** .
On page view: 
![image](https://user-images.githubusercontent.com/15880606/109560376-0bb61780-7a99-11eb-9f0c-1b9e8dfc1c85.png)

On admin view: 
![image](https://user-images.githubusercontent.com/15880606/109560532-3e601000-7a99-11eb-9e6d-5ecca3e00e35.png)



- [ ] [Issue] if I selected a dark bg, all the text inside will be white by default per this css: 

```
editor-styles-wrapper .uds-section.bg.morse-code-black .wp-block-wp-bootstrap-blocks-container, .editor-styles-wrapper .uds-section.bg.network-black .wp-block-wp-bootstrap-blocks-container, .editor-styles-wrapper .uds-section.bg.topo-white .wp-block-wp-bootstrap-blocks-container {
background-color: #191919;
    color: #fff;
}
```

In this case I created a **background section** with **dark** option, then inside that section, created another **background section** with **uploaded image** and **UDS White** option for the container background, but the text was still showing in white as it's a child of the first dark background section. 
I think this should only target the direct child. 
![image](https://user-images.githubusercontent.com/15880606/109562562-ca733700-7a9b-11eb-822f-a200f1908c88.png)

- [ ] [Issue] The topographic map - white and dark options need to be switched. 
![Screen Shot 2021-03-01 at 2 43 11 PM](https://user-images.githubusercontent.com/15880606/109563209-b24fe780-7a9c-11eb-82c5-290c871cce57.png)

SteveRyan-ASU commented:
Picking this back up and hopefully pushing it across the line. 

It's been a while since I last looked at this, but to get it back up to speed, I started with the reorganization of some of the ACF fields and simplified the controls a little bit.

The block still has three main "modes" which are triggered by the opening radio button for "pattern or color"
### UDS Background Texture ###
- Produces a background section with an approved pattern.
- Styles the inner container minimally and appropriately with either a white or UDS-Gray-7 color.
- Adds `text-white` class to the `<section>` element to assist with dark on dark text.

### ASU Background Color ###
- Presents a choice for a background color as of of the "official, supported" options: Gray 1, Gray 2 or Gray 7.
- Allows for a fourth choice to "choose my own color" with a sufficient warning about allowable colors from the palette.
- A color picker is presented to the user conditionally to allow for a custom color.

### Upload a file ###
- Presents a dialog for uploading a file into the media library.
- Conditionally presents an additional accordion of options pertaining to the CSS background style.
- Allows for a background-position and background-repeat option. Those two options handle the bulk of the difference between a photo background (no-repeat, cover) and a background pattern (repeat).
- Uploaded files also allow for the user to manually select an inner background color from the approved set of colors. (White, Gray 1, 2 and 7. Also allows for a no inner background color option.)

Addressing some of the concerns from @zainabD's last review:

- All styles applied to the inner "container" for this block are now targeting a class of [class*="container"] as opposed to just `.container`. The wildcard selector should cover all container types including fluid and responsive-fluid.
- Any style applied to the inner container from a class within the `section` can immediately be overridden by a custom CSS field applied to the container directly. 
- I added the utility code to allow for Additional CSS rules to be applied to this new block.
- The inner blocks feature of this block does not allow for a section to be nested within another section. So, via ordinary means, we shouldn't run into problems where the inner container rules need to be applied to only a direct child of the section block.

The opacity options that were initially within this block were dropped for simplicity. The best way to implement these back would be to allow the user to pick a custom inner container color via the color picker. (The output of the color picker can be presented as an RGBA value, so you get a color and an opacity in one shot.)

zainabD commented on css/admin.css:
```diff
@@ -91,10 +94,14 @@ Enqueued only within the admin screen. (Via enqueue.php)
   vertical-align: -.125rem; }
 
+<<<<<<< HEAD
+/*# sourceMappingURL=data:application/json;charset=utf8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImFkbWluLnNjc3MiLCJhZG1pbi5jc3MiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUE7Ozs7OzsrRENNK0Q7QURFL0Q7RUFDQyxnQkFBZSxFQUFBOztBQUdoQjs7cUVDQXFFO0FEV3BFO0VBRUMsc0JBUmtCLEVBQUE7RUFNbkI7SUFPRSx1QkFBdUIsRUFBQTtFQVB6QjtJQWFHLGlCQUFpQixFQUFBOztBQWJwQjtFQUVDLHNCQVJrQixFQUFBO0VBTW5CO0lBT0UsdUJBQXVCLEVBQUE7RUFQekI7SUFh ...
```
It looks like this wasn't merged correctly, why do we still have **<<<<<<< HEAD** and **>>>>>>> develop**

zainabD commented on css/admin.css:
```diff
@@ -91,10 +94,14 @@ Enqueued only within the admin screen. (Via enqueue.php)
   display: none; }
 
 /*# sourceMappingURL=data:application/json;charset=utf8;base64,eyJ2ZXJzaW9uIjozLCJzb3VyY2VzIjpbImFkbWluLnNjc3MiLCJhZG1pbi5jc3MiXSwibmFtZXMiOltdLCJtYXBwaW5ncyI6IkFBQUE7Ozs7OzsrRENNK0Q7QURFL0Q7O3FFQ0NxRTtBRFVwRTtFQUVDLHNCQVJrQixFQUFBO0VBTW5CO0lBT0UsdUJBQXVCLEVBQUE7RUFQekI7SUFhRyxpQkFBaUIsRUFBQTs7QUFicEI7RUFFQyxzQkFSa0IsRUFBQTtFQU1uQjtJQU9FLHVCQUF1QixFQUFBO0VBUHpCO0lBYUcsaUJBQWlCLEVBQUE7O0FBYnBCO0VB ...
+>>>>>>> develop
```
Same here ... 
I noticed this because I'm getting the following error on console: 
`Error while traversing the CSS: Error: undefined:7:1: missing '{'`
and the editor styling are broken

zainabD commented on css/admin.min.css:
```diff
@@ -1,2 +1,6 @@
+=======
 .menu-item-depth-3 .menu-item-bar .menu-item-handle{background-color:#f99}.menu-item-depth-3 .menu-item-bar .menu-item-handle .is-submenu::after{content:' - too deep!'}.menu-item-depth-3 .menu-item-bar .menu-item-handle .item-title{margin-right:5em}.menu-item-depth-4 .menu-item-bar .menu-item-handle{background-color:#f99}.menu-item-depth-4 .menu-item-bar .menu-item-handle .is-submenu::after{conte ...
+>>>>>>> develop
 /*# sourceMappingURL=admin.min.css.map */
```
same here, conflicts were not resolved when merged develop into this ..

zainabD commented on css/admin.min.css.map:
```diff
@@ -1 +1,5 @@
+{"version":3,"sources":["admin.scss"],"names":[],"mappings":"AAQA,WACC,WAAA,KAcA,oDAEC,iBAAA,KAFD,uEAOE,QAAA,eAPF,gEAaG,aAAA,IAbH,oDAEC,iBAAA,KAFD,uEAOE,QAAA,eAPF,gEAaG,aAAA,IAbH,oDAEC,iBAAA,KAFD,uEAOE,QAAA,eAPF,gEAaG,aAAA,IAbH,oDAEC,iBAAA,KAFD,uEAOE,QAAA,eAPF,gEAaG,aAAA,IAbH,oDAEC,iBAAA,KAFD,uEAOE,QAAA,eAPF,gEAaG,aAAA,IAbH,oDAEC,iBAAA,KAFD,uEAOE,QAAA,eAPF,gEAaG,aAAA,IAbH,oDAEC,iBAAA,KAFD,uEAOE,Q ...
+=======
+{"version":3,"sources":["admin.scss"],"names":[],"mappings":"AAmBC,oDAEC,iBAAA,KAFD,uEAOE,QAAA,eAPF,gEAaG,aAAA,IAbH,oDAEC,iBAAA,KAFD,uEAOE,QAAA,eAPF,gEAaG,aAAA,IAbH,oDAEC,iBAAA,KAFD,uEAOE,QAAA,eAPF,gEAaG,aAAA,IAbH,oDAEC,iBAAA,KAFD,uEAOE,QAAA,eAPF,gEAaG,aAAA,IAbH,oDAEC,iBAAA,KAFD,uEAOE,QAAA,eAPF,gEAaG,aAAA,IAbH,oDAEC,iBAAA,KAFD,uEAOE,QAAA,eAPF,gEAaG,aAAA,IAbH,oDAEC,iBAAA,KAFD,uEAOE,QAAA,eAPF,gEAaG ...
+>>>>>>> develop
```
Same issue

zainabD commented on css/theme.min.css:
```diff
@@ -1,2 +1,6 @@
+<<<<<<< HEAD
+@charset "UTF-8";:focus{outline:0;-webkit-box-shadow:0 0 8px #00baff!important;box-shadow:0 0 8px #00baff!important}body{font-family:Arial,Helvetica,"Nimbus Sans L","Liberation Sans",FreeSans,sans-serif;font-weight:400;color:#191919}.text-white{color:#e8e8e8!important}.text-underline{text-decoration:underline}.text-capitalize{text-transform:capitalize!important}ins,u{-webkit-text-decoration-line: ...
```
<<<<<<< HEAD and >>>>>>> develop

zainabD commented on inc/uds-blocks.php:
```diff
@@ -53,6 +53,7 @@ function my_acf_blocks_init() {
 			'/content-sections', // Miscellaneous content sections.
 			'/headings', // A UDS Headings block.
 			'/program-card', // UDS Program Cards.
+			'/background-section', // UDS Background section.
```
I added this line since the block was not registered anywhere.

zainabD commented on templates-blocks/background-section/background-section.php:
```diff
@@ -0,0 +1,74 @@
+if ( ! empty( $block['className'] ) ) {
+	$additional_classes = $block['className'];
+}
+
```
Can you give the $additional_classes an empty value if it's empty to avoid the following error: 
`Notice: Undefined variable: additional_classes`

SteveRyan-ASU replied:
Resolved. Good catch. :-)

zainabD reviewed (changes requested):
Over all it looks good, I just added a note to fix the error "`Notice: Undefined variable: additional_classes`"
- Also, I have a couple of suggestions: 

- [ ] In image background case, can we add an additional option beside the background-position and repeat, a new option to control the **background size** whether it's contain or cover? 

- [ ] Is it possible to add the option of curve backgrounds, like the ones show on SFIS and SOS xd mockup: 

![image](https://user-images.githubusercontent.com/15880606/115312216-c8396a80-a125-11eb-932f-a42b2e315a97.png)
 
We added these to SFIS and SOS home page an additional css as shown below (https://sfis.qa.rtd.asu.edu/): 

```
    border-bottom-left-radius: 55% 10%;
    border-bottom-right-radius: 55% 10%;
    border-top-left-radius: 55% 10%;
    border-top-right-radius: 55% 10%;
    width: 1900px;
    background-position: center;
    position: relative;
    left: 50%;
    transform: translateX(-50%);
```

zainabD commented:
One more thing I forgot to mention: 
In SFIS and SOS, we have use a background image that is only repeats on x, like this one: 
![image](https://user-images.githubusercontent.com/15880606/115317384-46027380-a130-11eb-8490-8c14d4096cbe.png)

Can we add this option under background repeat, to choose whether to repeat on x or on y?

SteveRyan-ASU commented:
Requested changes in the latest push to this branch:

- The choice of the inner text color is now a user-driven choice rather than something that was applied at the CSS level with the choice of "approved" background textures. The control defaults to normal text until you explicitly set it as white. 
- Added support for both repeat-x and repeat-y as `background-repeat` options. The options of `space` and `round` were omitted to create a better user experience, although adding them back would be trivial if desired.
- Added explicit support for the `background-size` options of cover and contain. I think it would be possible to add a conditional control within the block that would allow the user to define a string for an explicit numeric value (`30% auto`, for example). But, that is also possible to do by adding a custom CSS class and rolling the rule out that way.

Adding these choices as completely user-configurable options DOES allow the block to produce some code that wouldn't normally make sense. (`background-size:cover` does not require an explicit `no-repeat` value, for example.) But, the produced code is still valid markup

I also added the oval background shape as a utility class to this element, but I left the feature without an explicit control within the UI to turn it on or off. 
- Add the class of `.shape-oval` to the advanced options area to trigger the result.
- Unfortunately, with this method of delivering this feature, the result is not visible within Gutenberg.
- Plays well with all of the other options within the block including the "roll your own background image" and background colors. Also respects the `top/middle/bottom` rules of the background-position controls within the block.

zainabD reviewed (approved):
Nice and clean work! good job Steve.
I tested this locally and everything seems to be working with no issues.

wmcconnell reviewed (commented):
I don't know if it was just me, but the custom background image was not working as I expected. I chose an image (a repeatable pattern, 240px wide) as my background and got this:

![Screen Shot 2021-05-06 at 3 28 37 PM](https://user-images.githubusercontent.com/31013359/117373319-dd313000-ae7f-11eb-892e-daea98556560.png)

The small image is stretched, and not repeating. The final CSS seems to have an unwanted character in there, a `\` between `center` and `cover`:

```<section class="uds-section media-file inner-bg-white " style="background: url(http://mythemedev.local/wp-content/uploads/2021/05/wildtextures_brick-elevation-tiles-seamless-texture-e1498137214248-480x461-1.jpg) center / cover repeat;">```

I'm not even sure if you want both `center` and `cover`. Removing `/ cover` fixes the issue, with the image actually repeating and looking like this:

![Screen Shot 2021-05-06 at 3 32 06 PM](https://user-images.githubusercontent.com/31013359/117373582-5d579580-ae80-11eb-937c-686b954eca0d.png)

Haven't figured out where the `/` comes from, but could it be taking the place of an actual parameter and causing some issues with the repeat setting?

SteveRyan-ASU commented:
Good catch @wmcconnell. Lucky for me the fix was small. 

- Added "auto" as a background-size option within ACF and set it as the default.

**Details**

The `/` within the `background:` inline CSS rule is part of the spec. It's a required value when setting both a background position and size when using the shorthand rule set. (https://developer.mozilla.org/en-US/docs/Web/CSS/background)

Removing the `/ cover` or `/ contain` declarations from the rule still produces valid markup that is the equivalent of `/ auto` for the background-size. 

I didn't catch the difference visually between "contain" and "auto" because the tile I was using to test was too small to notice the difference. Here's the difference:

- Both "contain" and "auto" when used with either of the repeat settings will produce a tiled background image.
- Contain will have the effect of scaling down the tiles so that none of the tiles are cut off from the size of the background area. (It will repeat an exact whole number of times within the given area.)
- The "auto" value does what you would expect - repeats and cuts off the excess of the image based on the background-position.

wmcconnell reviewed (approved):
Nice work, Steve! Easy to create a series of page sections with various background styles and with the proper UDS spacing built in. I was able to quickly fill a page with multiple sections containing images, text, and cards and have it look great with minimal effort.
